src/main/java/sqlline/SqlLinePropertiesEnum.java

```diff
@@ -0,0 +1,119 @@
+/*
+// Licensed to Julian Hyde under one or more contributor license
+// agreements. See the NOTICE file distributed with this work for
+// additional information regarding copyright ownership.
+//
+// Julian Hyde licenses this file to you under the Modified BSD License
+// (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at:
+//
+// http://opensource.org/licenses/BSD-3-Clause
+*/
+package sqlline;
+
+import java.io.File;
+
+/**
+ * Properties that may be specified for SqlLine.
+ */
+public enum SqlLinePropertiesEnum implements SqlLineProperty {
+
+  AUTO_COMMIT("autoCommit", Type.BOOLEAN, true),
+  AUTO_SAVE("autoSave", Type.BOOLEAN, false),
+  COLOR("color", Type.BOOLEAN, false),
+  CSV_DELIMITER("csvDelimiter", Type.STRING, ","),
+
+  CSV_QUOTE_CHARACTER("csvQuoteCharacter", Type.CHAR, '\''),
+
+  DATE_FORMAT("dateFormat", Type.STRING, DEFAULT),
+
+  FAST_CONNECT("fastConnect", Type.BOOLEAN, true),
+  FORCE("force", Type.BOOLEAN, false),
+  HEADER_INTERVAL("headerInterval", Type.INTEGER, 100),
+  HISTORY_FILE("historyFile", Type.STRING,
+      new File(SqlLineOpts.saveDir(), "history").getAbsolutePath()),
+  INCREMENTAL("incremental", Type.BOOLEAN, true),
+  ISOLATION("isolation", Type.STRING, "TRANSACTION_REPEATABLE_READ"),
+  MAX_COLUMN_WIDTH("maxColumnWidth", Type.INTEGER, 15),
+  // don't save maxheight, maxwidth: it is automatically set based on
+  // the terminal configuration
+  MAX_HEIGHT("maxHeight", Type.INTEGER, 80, false, false),
+  MAX_WIDTH("maxWidth", Type.INTEGER, 80, false, false),
+
+  NUMBER_FORMAT("numberFormat", Type.STRING, DEFAULT),
+  NULL_VALUE("nullValue", Type.STRING, DEFAULT),
+  SILENT("silent", Type.BOOLEAN, false),
+  OUTPUT_FORMAT("outputFormat", Type.STRING, "table"),
+  ROW_LIMIT("rowLimit", Type.INTEGER, 0),
+  SHOW_ELAPSED_TIME("showElapsedTime", Type.BOOLEAN, true),
+  SHOW_HEADER("showHeader", Type.BOOLEAN, true),
+  SHOW_NESTED_ERRS("showNestedErrs", Type.BOOLEAN, false),
+  SHOW_WARNINGS("showWarnings", Type.BOOLEAN, true),
+  TIME_FORMAT("timeFormat", Type.STRING, DEFAULT),
+  TIMEOUT("timeout", Type.INTEGER, -1),
+  TIMESTAMP_FORMAT("timestampFormat", Type.STRING, DEFAULT),
+  TRIM_SCRIPTS("trimScripts", Type.BOOLEAN, true),
+  VERBOSE("verbose", Type.BOOLEAN, false),
+  VERSION("version", Type.STRING, new Application().getVersion(), false, true);
+
+  private final String propertyName;
+  private final Type type;
+  private final Object defaultValue;
+  private final boolean couldBeStored;
+  private final boolean isReadOnly;
+
+  SqlLinePropertiesEnum(String propertyName, Type type, Object defaultValue) {
```

Comment:
Should property also store its description?

Reply:
Not sure about description as currently it is located in `SqlLine.properties` and 
1. potentially resource for another locale could be added
2. could be changed without code changes

in case of putting description into code for both points the code changes will be required

Reply:
Agree about the locale, it's just you always have to remember to add property description in the SqlLine.properties file, which developers tend to forget. Anyway, I don't mind if we leave as is, just wanted to bring up this point.